From e1fd5485e1956ecb0eecf662c3260d01ae04b11b Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Mon, 29 May 2023 15:56:39 +0800
Subject: [PATCH 47/53] [Huawei] mm/sharepool: Delete is_alive element in
 sp_group structure

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Since we have merged sp_group_exit and sp_group_post_exit, the dead
state for a sp_group would never exist. Only use the use_count to
manage its lifetime.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 41 +++--------------------------------------
 1 file changed, 3 insertions(+), 38 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 152042ee18af..285997242921 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -51,8 +51,6 @@
 #include <linux/time64.h>
 #include <linux/pagewalk.h>
 
-#define spg_valid(spg)		((spg)->is_alive == true)
-
 /* Use spa va address as mmap offset. This can work because spa_file
  * is setup with 64-bit address space. So va shall be well covered.
  */
@@ -172,8 +170,6 @@ struct sp_group {
 	struct rb_root	spa_root;
 	/* group statistics */
 	struct sp_meminfo meminfo;
-	/* is_alive == false means it's being destroyed */
-	bool		is_alive;
 	atomic_t	use_count;
 	atomic_t	spa_num;
 	/* protect the group internal elements */
@@ -1058,7 +1054,6 @@ static bool is_online_node_id(int node_id)
 static void sp_group_init(struct sp_group *spg, int spg_id)
 {
 	spg->id = spg_id;
-	spg->is_alive = true;
 	spg->proc_num = 0;
 	spg->spa_root = RB_ROOT;
 	atomic_set(&spg->use_count, 1);
@@ -1162,17 +1157,8 @@ static struct sp_group *sp_group_get_or_alloc(int spg_id)
 	struct sp_group *spg;
 
 	spg = idr_find(&sp_group_idr, spg_id);
-	if (!spg || !atomic_inc_not_zero(&spg->use_count)) {
+	if (!spg || !atomic_inc_not_zero(&spg->use_count))
 		spg = sp_group_create(spg_id);
-	} else {
-		down_read(&spg->rw_lock);
-		if (!spg_valid(spg)) {
-			up_read(&spg->rw_lock);
-			sp_group_put_locked(spg);
-			return ERR_PTR(-ENODEV);
-		}
-		up_read(&spg->rw_lock);
-	}
 
 	return spg;
 }
@@ -1893,14 +1879,6 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 		goto drop_spa;
 	}
 
-	if (!spg_valid(spg)) {
-		up_write(&spg->rw_lock);
-		if (spa->type != SPA_TYPE_ALLOC)
-			vmalloc_area_clr_flag(spa->kva, VM_SHAREPOOL);
-		goto drop_spa;
-		/* we must return success(0) in this situation */
-	}
-
 	__sp_free(spa, NULL);
 	if (spa->type == SPA_TYPE_ALLOC) {
 		sp_fallocate(spa);
@@ -2236,12 +2214,6 @@ void *__mg_sp_alloc_nodemask(unsigned long size, unsigned long sp_flags, int spg
 	}
 
 	down_write(&spg->rw_lock);
-	if (!spg_valid(spg)) {
-		up_write(&spg->rw_lock);
-		sp_group_put(spg);
-		pr_info("allocation failed, spg is dead\n");
-		return ERR_PTR(-EINVAL);
-	}
 try_again:
 	spa = sp_area_alloc(ac.size_aligned, ac.sp_flags, spg,
 			    ac.type, current->tgid, ac.preferred_node_id);
@@ -3231,15 +3203,8 @@ static void spa_stat_of_mapping_show(struct seq_file *seq, struct sp_mapping *sp
 	for (node = rb_first(&spm->area_root); node; node = rb_next(node)) {
 		spa = rb_entry(node, struct sp_area, rb_node);
 
-		if (spg_valid(spa->spg))  /* k2u to group */
-			seq_printf(seq, "%-10d ", spa->spg->id);
-		else  /* spg is dead */
-			seq_printf(seq, "%-10s ", "Dead");
-
-		seq_printf(seq, "%2s%-14lx %2s%-14lx %-10ld ",
-			   "0x", spa->va_start,
-			   "0x", spa->va_end,
-			   byte2kb(spa->real_size));
+		seq_printf(seq, "%-10d 0x%-14lx 0x%-14lx %-10ld ",
+			   spa->spg->id, spa->va_start, spa->va_end, byte2kb(spa->real_size));
 
 		switch (spa->type) {
 		case SPA_TYPE_ALLOC:
-- 
2.17.1

